server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -426,16 +476,30 @@ public DocumentField field(String fieldName) {
      * were required to be loaded.
      */
     public Map<String, DocumentField> getFields() {
-        return fields == null ? emptyMap() : fields;
+        Map<String, DocumentField> fields = new HashMap<>();
+        fields.putAll(metaFields);
+        fields.putAll(documentFields);
+        return fields;
     }
 
     // returns the fields without handling null cases
     public Map<String, DocumentField> fieldsOrNull() {
-        return fields;
+        return getFields();
     }
 
     public void fields(Map<String, DocumentField> fields) {
-        this.fields = fields;
+        this.metaFields.clear();
```

Comment:
Why clear? We should just set to maps.

Reply:
That was a way to make sure that the new content coming from `fields` map does not get mixed with old content of the maps. If I got you right, this will be replaced with creation of 2 new empty hash maps here.